promise
===================
Promise对象，可以将异步操作以同步操作的流程表达出来，避免了层层嵌套的回调函数。
###用法
```
const promise = new Promise(function(resolve, reject) {
  // code
});
```
Promise构造函数接受一个函数作为参数，这个参函数也有两个参数resolve和reject，它们也是函数。resolve函数将Promise对象的状态从“未完成”变为“成功”，所以在异步操作成功时调用；reject函数将Promise对象的状态从“未完成”变为“失败”，所以在异步操作失败时调用，这两个函数都可以传递参数。
注意：**Promise 创建后就会立即执行**。所以，用Promise的时候一般是包在一个函数中，在需要的时候去运行这个函数，如：
```
function runPromise(){
    const promise = new Promise(function(resolve, reject) {
      // code
    });
    return promise;
}
```
then方法是Promise对象“成功”和“失败”的回调函数。有两个函数作为参数，这两个参函数分别接受resolved（成功）状态和rejected（失败）状态传递的值为参数：
```
promise.then(function(value) {
  // 异步成功后需要执行的代码
}, function(error) {
  // 异步失败后需要执行的代码
});
```
第二个函数是可选的，不一定要提供。

下面是异步加载图片的例子。
```
function loadImageAsync(url) {
  return new Promise(function(resolve, reject) {
    const image = new Image();

    image.onload = function() {
      resolve(image);
    };

    image.onerror = function() {
      reject(new Error('Could not load image at ' + url));
    };

    image.src = url;
  });
}

loadImageAsync("imgSrc").then(function(value){
  //异步成功的回调
  document.body.appendChild(value)
},function(error){
  //异步失败的回调
  console.error(error)
})
```
###链式调用
then支持执行链式调用：
```
doSomething().then(function(result) {
  return doSomethingElse(result);
})
.then(function(newResult) {
  return doThirdThing(newResult);
})
.catch(failureCallback);
```
如果前一个then，返回值还是一个Promise对象，那么后一个then，就会等待该Promise对象的状态发生变化，才会被调用。
如果返回的同步操作，则直接调用后一个then。

###catch方法
Promise对象还有一个catch方法来收集错误：

	promise.then(function(data){
		console.log(data);
	}).catch(function(error){
		console.log(error)
	})

Promise 对象的错误具有“冒泡”性质，会一直向后传递，直到被捕获为止。所以，只要在最后添加catch方法就可以，不需要reject回调。

	loadImageAsync("imgSrc1").then((image)=>{
		document.body.appendChild(image);
		// 注意，then()执行后需要返回新的promise用于链式调用
		return loadImageAsync("imgSrc2")
	}).then((image)=>{
		document.body.appendChild(image);
	}).catch((error)={
		console.log(error)
	})
Promise 内部的错误会终止后续所有then的运行，并传递到catch内，但是不会影响到 Promise 外部的代码。

###all方法
all()方法将多个 Promise 实例合成一个新的 Promise 实例。
```
let p = Promise.all([p1, p2, p3]);
```
all()方法接收一个具有 Iterator 接口的对象为参数，上例p的状态由p1、p2、p3决定，p1、p2、p3都成功，则p变为成功，p1、p2、p3有一个失败，则p失败。

###race方法
race()方法也是将多个 Promise 实例合成一个新的 Promise 实例。
```
let p = Promise.race([p1, p2, p3]);
```
上例p的状态由p1、p2、p3决定，但不同于all方法，p1、p2、p3谁先改变状态，p就相应改变状态。
```
const p = Promise.race([
  fetch('/resource-that-may-take-a-while'),
  new Promise(function (resolve, reject) {
    setTimeout(() => reject(new Error('request timeout')), 5000)
  })
]);

p
.then(console.log)
.catch(console.error);
```
上面代码中，如果 5 秒之内fetch方法无法返回结果，就触发rejected。
